### PR TITLE
[Spec] Correctly validate payeeOrigin

### LINKED ISSUE #149: What is a fully qualified origin?
I came across

> If request.payeeOrigin is not a fully qualified origin, return false.

and it's not at all clear to me how this works. Are you parsing the input? How?

### PR DESCRIPTION
**Pull request by stephenmcgruer**
Fixes #149

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/secure-payment-confirmation/pull/160.html" title="Last updated on Nov 15, 2021, 5:07 PM UTC (8a23d12)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/secure-payment-confirmation/160/8fbbe76...8a23d12.html" title="Last updated on Nov 15, 2021, 5:07 PM UTC (8a23d12)">Diff</a>

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
@annevk - PTAL (and apologies for the delay!). I wasn't sure if we should be doing the same 3rd step as [postMessage](https://html.spec.whatwg.org/multipage/web-messaging.html#window-post-message-steps) does, which is:

`Set targetOrigin to parsedURL's origin.`

I assume this is for normalization, but I'm unclear if its desirable for SPC as it seems to be a tuple rather than a string then?

**Comment by annevk** [comment]
If you don't have an actual origin, how would you do a same origin check? Internally you also want origins to be "objects". If you need to output it at some point you can serialize it then.

**Comment by stephenmcgruer** [comment]
> how would you do a same origin check

I may not be following you; nowhere in the spec is `payeeOrigin` subject to a same-origin check, I believe. The `payeeOrigin` is a claim by the SPC caller as to which origin is receiving payment, which is displayed to the user by the browser and signed into the cryptogram. It may or may not be the same as the origin that SPC was called from.

For example, in a redirect flow, `merchant.com` will redirect to `psp.com`, who may then call SPC with `payeeOrigin: 'https://merchant.com'`. In this flow, the browser has no way to enforce that `payeeOrigin` is anything in particular, but that's ok - we don't have the ability to enforce that (e.g.) the transaction `amount` is correct either. The browser's job is to clearly indicate to the user what transaction information they are agreeing to, and then sign that for later validation/verification by the Relying Party (e.g., the bank).

**Comment by annevk** [comment]
Okay, so you essentially use a serialized form of the origin for signing (which would still end up being compared I suspect at some other point by another party?).

In that case you do still want to extract the origin from URL. Then you probably want to check that it's not an opaque origin (and fail if it is one). And I guess you might even want to enforce that the origin's scheme is `https`? Then you want to serialize the origin and use the resulting value for signing.

And yeah, normalization is important here. E.g., if the supplied value is `https://EXAMPLE.example/test`, you want to sign with `https://example.example`, right?

**Comment by stephenmcgruer** [comment]
Ah yes, very good points. I'm embarrassed - I hadn't even considered that (I now assume) the URL parser would accept `https://foo.com/this/is/a/path`, nevermind things like opaque origins >_<.

Thanks, I'll update the PR and ping for re-review. I appreciate your help!

**Comment by stephenmcgruer** [comment]
@annevk - PTAAL. I now check the `scheme` is https (which I believe rules out opaque origins as well), and then overwrite the input with the normalization. It's a little dodgy to do the overwrite (since it makes the `Steps to check if a payment can be made` non-idempotent), but in practice it 'works' and long-term I expect us to move SPC away from PaymentRequest and to its own API shape (see https://github.com/w3c/secure-payment-confirmation/issues/65#issuecomment-902115127)

**Review by annevk — COMMENTED** [review]
Yeah this works I think. Modifying _data_ is a little weird, but it's a copy so it should be okay.

File: spec.bs [hunk 1]
```diff
@@ -60,6 +60,12 @@ spec: html; urlPrefix: https://html.spec.whatwg.org/multipage/
     type: dfn
         urlPrefix: origin.html
             text: origin; url: concept-origin
+            text: scheme; url: concept-origin-scheme
+            text: serialization of an origin; url: ascii-serialisation-of-an-origin
+
+spec: url; urlPrefix: https://url.spec.whatwg.org/
+    type: dfn
+        text: URL parser; url: concept-url-parser
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply spec.bs]: Done.

File: spec.bs [hunk 2]
```diff
@@ -504,8 +510,8 @@ members:
         details times out. At most 1 hour.
 
     :  <dfn>payeeOrigin</dfn> member
-    :: The fully qualified [=origin=] of the payee that this SPC call is for
-        (e.g., the merchant).
+    :: The [=origin=] of the payee that this SPC call is for (e.g., the
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply spec.bs]: Done.

File: spec.bs [hunk 3]
```diff
@@ -522,8 +528,15 @@ input {{SecurePaymentConfirmationRequest}} |data|, are:
...
     return `false`.
 
-1. If |data|["{{SecurePaymentConfirmationRequest/payeeOrigin}}"] is not a fully
-    qualified [=origin=], return `false`.
+1. Let |parsedURL| be the result of running the [=URL parser=] on
+    |data|["{{SecurePaymentConfirmationRequest/payeeOrigin}}"].
+
+1. If |parsedURL| is failure, then return `false`.
+
+1. If |parsedURL|'s [=scheme=] is not "https", return `false`.
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply spec.bs]: Done.

File: spec.bs [hunk 4]
```diff
@@ -522,8 +528,15 @@ input {{SecurePaymentConfirmationRequest}} |data|, are:
...
+
+1. Set |data|["{{SecurePaymentConfirmationRequest/payeeOrigin}}"] to the
+    [=serialization of an origin|serialization of=] |parsedURL|'s [=origin=].
```
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply spec.bs]: Done.

File: spec.bs [hunk 5]
```diff
@@ -522,8 +517,15 @@ input {{SecurePaymentConfirmationRequest}} |data|, are:
...
     return `false`.
 
-1. If |data|["{{SecurePaymentConfirmationRequest/payeeOrigin}}"] is not a fully
-    qualified [=origin=], return `false`.
+1. Let |parsedURL| be the result of running the [=URL parser=] on
+    |data|["{{SecurePaymentConfirmationRequest/payeeOrigin}}"].
+
+1. If |parsedURL| is failure, then return `false`.
+
+1. If |parsedURL|'s [=url/scheme=] is not "https", then return `false`.
```
Comment (annevk) [inline spec.bs:525]: nit
```suggestion
1. If |parsedURL|'s [=url/scheme=] is not "`https`", then return `false`.
```

**Review by annevk — APPROVED** [review]
This looks pretty solid, thanks.

(By the way, if you ever need to add something from a WHATWG specification to an anchors block, please file an issue. To a large extent they are code smell.)

**Comment by stephenmcgruer** [comment]
> (By the way, if you ever need to add something from a WHATWG specification to an anchors block, please file an issue. To a large extent they are code smell.)

Ack, thanks (and thanks for the review!)

**Pull request merged**